### PR TITLE
types: blob - extract createcommitment

### LINKED ISSUE #25: types: Blob API needs CreateCommitment
To be able to submit blobs using Blob module API, we need to pass a `Commitment` which can be created from `CreateCommitment`. This method is missing from types. This is a blocker for using Blob  API in OP stack to replace the frame pointer to use a commitment instead of block height + tx index.

### PR DESCRIPTION
**Pull request by tuxcanfly**
## Overview

This PR extracts `CreateCommitment` from `celestia-app`. This can be used to submit blob commitments with minimal dependencies.

Fixes #25 

## Checklist

- [x] New and updated code has appropriate documentation
- [x] New and updated code has new and/or updated testing
- [x] Required CI checks are passing
- [x] Visual proof for any user facing features like CLI or documentation updates
- [x] Linked issues closed with keywords

### CODE REVIEW

**Comment by tuxcanfly** [comment]
There seems to be an issue that happens only when passing commitment:

`unsupported namespace id length: id [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 2 3 4 5 6 7 8] must be 28 bytes but it was 29 bytes`

**Comment by gupadhyaya** [comment]
@tuxcanfly can you split the code in types/blob/commitment.go to multiple files? its too large and not a good idea to dump everything to one file. you can retain the structure/names as in celestia-app or make up your own meaningful structure/names.

**Comment by tuxcanfly** [comment]
@gupadhyaya Done. Also modified `CreateCommitment` and dependencies to use the combined `Namespace` id and version bytes to ease confusion. This means that any code that's copied from `celestia-app` in the future needs to be modified to use the full `Namespace` bytes e.g. in `SpareShareSplitter.Write`:

```
appns.New(blob.NamespaceVersion, blob.NamespaceID)
```

becomes

```
namespace.From(blob.Namespace)
```

As a reminder, `celestia-core/types/block.go` defines separate version and id bytes:

```
type Blob struct {
	// NamespaceVersion is the version of the namespace. Used in conjunction
	// with NamespaceID to determine the namespace of this blob.
	NamespaceVersion uint8

	// NamespaceID defines the namespace ID of this blob. Used in conjunction
	// with NamespaceVersion to determine the namespace of this blob.
	NamespaceID []byte

	// Data is the actual data of the blob.
	// (e.g. a block of a virtual sidechain).
	Data []byte

	// ShareVersion is the version of the share format that this blob should use
	// when encoded into shares.
	ShareVersion uint8
}
```

while `celestia-node/blob/blob.go` defines:

```
type Blob struct {
	types.Blob `json:"blob"`

	Commitment Commitment `json:"commitment"`

	// the celestia-node's namespace type
	// this is to avoid converting to and from app's type
	namespace share.Namespace
}
```

where `share.Namespace` contains combined version and id bytes:

```
// Namespace represents namespace of a Share.
// Consists of version byte and namespace ID.
type Namespace []byte
```

Since we need to pass the combined `Namespace` bytes in the rpc client. It makes sense for us to _always_ use this and not have to maintain two different structs.

tldr: `types/blob/blob.go:Blob.Namespace` _must_ always contain both id and version bytes and references from `CreateCommitment` should always be updated to handle this.

**Comment by nashqueue** [comment]
cc @rootulp @evan-forbes

File: types/blob/share_builder.go [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+			return nil, err
+		}
+	} else {
+		if err := b.prepareSparseShare(); err != nil {
+			return nil, err
+		}
+	}
+
+	return b, nil
+}
```
Comment (MSevey) [inline types/blob/share_builder.go:48]: Why are we returning the `Builder`? It isn't a new builder and the methods are just modifying the original?

Can't we just do this?

```suggestion
func (b *Builder) Init() (error) {
	if b.isCompactShare {
		return b.prepareCompactShare()
	} 
	return b.prepareSparseShare()
}
```

**Review by MSevey — DISMISSED** [review]
LGTM: just resolve any of my previous comments once the corresponding issues have been created in app

File: types/blob/share_builder.go (hunk 1 above)
Comment (tuxcanfly) [reply types/blob/share_builder.go:48]: Follow up: https://github.com/celestiaorg/celestia-app/issues/2157

File: types/appconsts/versioned_consts.go [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+// roots in the share commitment. If a blob contains more shares than this
+// number, then the height of the subtree roots will increase by one so that the
+// number of subtree roots in the share commitment decreases by a factor of two.
+// This step is repeated until the number of subtree roots is less than the
+// SubtreeRootThreshold.
+//
+// The rationale for this value is described in more detail in ADR-013.
+func SubtreeRootThreshold(_ uint64) int {
+	return v1.SubtreeRootThreshold
+}
```
[the remark this answers is not in the document]
Comment (tuxcanfly) [reply types/appconsts/versioned_consts.go:19]: follow up: https://github.com/celestiaorg/celestia-app/issues/2158

File: types/share/share.go [hunk 3]
```diff
@@ -1,14 +1,66 @@
...
+		return 0, fmt.Errorf("byteIndex must be less than share size %d", appconsts.ShareSize)
+	}
+	return byteIndex, nil
+}
+
+// InfoByte is a byte with the following structure: the first 7 bits are
+// reserved for version information in big endian form (initially `0000000`).
+// The last bit is a "sequence start indicator", that is `1` if this is the
+// first share of a sequence and `0` if this is a continuation share.
+type InfoByte byte
```
Comment (gupadhyaya) [inline types/share/share.go:32]: why are these repeated in here and also in info_byte.go?

File: types/blob/info_byte.go
Comment (gupadhyaya) [inline types/blob/info_byte.go:1]: recommend moving all the share related code to share package instead of blob package.

Comment (gupadhyaya) [reply types/blob/info_byte.go:1]: similar to what celestia-app has. e.g., info_byte, blob_share_commitment_rules, padding, powers_of_two, reserved_bytes, share_builder, share_splitting, split_sparse_shares and utils

Comment (gupadhyaya) [reply types/blob/info_byte.go:1]: blob/
 - blob.go
 - commitment.go
 - merkle.go
 
 share/
  - (rest of the files)

Comment (tuxcanfly) [reply types/blob/info_byte.go:1]: This was creating a circular dependency as `blob` depends on `share`: `func NewBlobV0(namespace share.Namespace, data []byte)` and `share` depends on `blob`: `func (sss *SparseShareSplitter) Write(blob Blob)`.

File: types/share/share.go (hunk 3 above)
Comment (tuxcanfly) [reply types/share/share.go:32]: Moved to `pkg share` 👍🏽

File: types/blob/info_byte.go
Comment (gupadhyaya) [reply types/blob/info_byte.go:1]: you can move Namespace to `namespace` pkg?

Comment (tuxcanfly) [reply types/blob/info_byte.go:1]: Unfortunately `share.Namespace` is different from `namespace.Namespace` which already exists and is confusing as it is 😅

Comment (tuxcanfly) [reply types/blob/info_byte.go:1]: We can try unifying these types here maybe... let me give it a try.

**Review by evan-forbes — COMMENTED** [review]
maybe we could add a readme that has a brief explanation of why the code is currently here, and also points to the original code. That might help the folks that are source diving and trying to link this code to the rest of what is going on.

**Review by evan-forbes — APPROVED** [review]
I'll defer to other reviewers, but I didn't see anything weird or out of the ordinary :slightly_smiling_face:

File: types/blob/info_byte.go
Comment (tuxcanfly) [reply types/blob/info_byte.go:1]: Moved almost everything over to `share` except `SplitBlobs` which needs to pass `...Blob` and would require another struct to hold the blob slice. Think it is reasonable to leave it out for the moment.

Comment (tuxcanfly) [reply types/blob/info_byte.go:1]: The file is `types/blob/share_splitting.go` so the structure now is:

```
tree types/blob
types/blob
├── blob.go
├── blob_test.go
├── commitment.go
├── merkle.go
└── share_splitting.go

tree types/share
types/share
├── blob_share_commitment_rules.go
├── errors.go
├── info_byte.go
├── namespace.go
├── padding.go
├── powers_of_two.go
├── reserved_bytes.go
├── share.go
├── share_builder.go
├── split_sparse_shares.go
└── utils.go
```

also added unit test for blob marshal/unmarshal.

Comment (tuxcanfly) [reply types/blob/info_byte.go:1]: Also, there wasn't an issue with `namespace` as much as passing `blob` to share splitter.

**Review by gupadhyaya — APPROVED** [review]

**Review by nashqueue — APPROVED** [review]
Thank you !
